fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,407 @@
+     * @param ioContext the IOContext from Lucene
+     * @return IndexOutput FDB Backed Index Output FDBIndexOutput
+     */
+    @Override
+    @Nonnull
+    public IndexOutput createOutput(@Nonnull final String name, @Nullable final IOContext ioContext) {
+        LOGGER.trace("createOutput -> {}", name);
+        return new FDBIndexOutput(name, name, this);
+    }
+
+    /**
+     * Lucene uses temporary files internally that are actually serialized onto disk.  Once on disk,
+     * the files can be renamed.
+     *
+     * @param prefix prefix
+     * @param suffix suffix
+     * @param ioContext ioContext
+     * @return IndexOutput
+     */
+    @Override
+    @Nonnull
+    public IndexOutput createTempOutput(@Nonnull final String prefix, @Nonnull final String suffix, @Nonnull final IOContext ioContext) {
+        LOGGER.trace("createTempOutput -> prefix={}, suffix={}", prefix, suffix);
+        return createOutput(getTempFileName(prefix, suffix, this.nextTempFileCounter.getAndIncrement()), ioContext);
+    }
+
+    @Nonnull
+    protected static String getTempFileName(@Nonnull String prefix, @Nonnull String suffix, long counter) {
+        return IndexFileNames.segmentFileName(prefix, suffix + "_" + Long.toString(counter, 36), "tmp");
+    }
+
+    @Override
+    public void sync(@Nonnull final Collection<String> collection) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("sync -> {}", String.join(", ", collection));
+        }
+    }
+
+    @Override
+    public void syncMetaData() throws IOException {
+        LOGGER.trace("syncMetaData");
+    }
+
+    /**
+     * It is the caller's responsibility to make the dest (destination) does not exist.
+     *
+     * @param source source
+     * @param dest desc
+     */
+    @Override
+    public void rename(@Nonnull final String source, @Nonnull final String dest) {
+        LOGGER.trace("rename -> source={}, dest={}", source, dest);
+        final byte[] key = metaSubspace.pack(source);
+        txn.get(key).thenAcceptAsync( (value) -> {
+            this.fileReferenceCache.invalidate(source);
+            txn.set(metaSubspace.pack(dest), value);
+            txn.clear(key);
+        }).join();
+    }
+
+    @Override
+    @Nonnull
+    public IndexInput openInput(@Nonnull final String name, @Nonnull final IOContext ioContext) throws IOException {
+        LOGGER.trace("openInput -> name={}", name);
+        return new FDBIndexInput(name, this);
+    }
+
+    @Override
+    @Nonnull
+    public Lock obtainLock(@Nonnull final String lockName) throws IOException {
+        LOGGER.trace("obtainLock -> {}", lockName);
+        return lockFactory.obtainLock(null, lockName);
+    }
+
+    /**
+     * No Op that reports in debug mode the block and file reference cache stats.
+     */
+    @Override
+    public void close() {
+        LOGGER.debug("close called blockCacheStats={}, referenceCacheStats={}", blockCache.stats(), fileReferenceCache.stats());
```

Comment:
Followup: We should be using `KeyValueLogMessage` for logging, at least for everything above trace. Automated logging ingesters appreciate standardized formats, and `KeyValueLogMessage` helps make sure at least Record Layer is consistent.